agent/src/util/device_plugin_builder.rs

```diff
@@ -0,0 +1,202 @@
+use super::{
+    constants::{K8S_DEVICE_PLUGIN_VERSION, KUBELET_SOCKET},
+    device_plugin_service::{DevicePluginService, InstanceMap},
+    v1beta1,
+    v1beta1::{device_plugin_server::DevicePluginServer, registration_client, DevicePluginOptions},
+};
+use akri_discovery_utils::discovery::v0::Device;
+use akri_shared::{
+    akri::{configuration::KubeAkriConfig, AKRI_PREFIX},
+    uds::unix_stream,
+};
+use async_trait::async_trait;
+use futures::stream::TryStreamExt;
+use log::{info, trace};
+#[cfg(test)]
+use mockall::{automock, predicate::*};
+use std::{convert::TryFrom, env, path::Path, time::SystemTime};
+use tokio::{
+    net::UnixListener,
+    net::UnixStream,
+    sync::{broadcast, mpsc},
+    task,
+};
+use tonic::transport::{Endpoint, Server, Uri};
+use tower::service_fn;
+
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait DevicePluginBuilderInterface: Send + Sync {
+    async fn build_device_plugin(
+        &self,
+        instance_name: String,
+        config: &KubeAkriConfig,
+        shared: bool,
+        instance_map: InstanceMap,
+        device_plugin_path: &str,
+        device: Device,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>>;
+}
+
+/// For each Instance, builds a Device Plugin, registers it with the kubelet, and serves it over UDS.
+pub struct DevicePluginBuilder {}
+
+#[async_trait]
+impl DevicePluginBuilderInterface for DevicePluginBuilder {
+    /// This creates a new DevicePluginService for an instance and registers it with the kubelet
+    async fn build_device_plugin(
```

Comment:
These methods have been moved from `device_plugin_service.rs` into a Trait so as to enable mocking and testing. They have not been modified.